### PR TITLE
Transform to json

### PR DESCRIPTION
**Pull request by sambolino**
A key update.

Contains functionality to serialize json dictionaries of alerts, as requested by Maria Patterson.
Both VOEvent and json historical calculations are now done with lc generator.

### CODE REVIEW

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (danielsf) [inline python/lsst/sims/alertsim/alertsim_main.py]: I think you need to update the docstring of `query_and_dispatch`.  You say that `obs_data` is the data from a catsim query, but at line 199, you treat it like an `InstanceCatalog` instance (i.e. you call `obs_data.iter_catalog()`, which makes me think `obs_data` is an instantiation of an `InstanceCatalog` class).

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (danielsf) [inline python/lsst/sims/alertsim/alertsim_main.py]: You will need to make this
```
ra1=obs_metadata.pointingRA - radius*np.cos(np.radians(obs_metadata.pointingDec))
ra2=obs_metadata.pointingDec + radius*np.cos(np.radians(obs_metadata.pointingDec))
```
to account for the fact that a change of 1 degree in RA means a different absolute angular distance at the poles than at the equator

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (danielsf) [inline python/lsst/sims/alertsim/alertsim_main.py]: What is the significance of `12435`?

File: python/lsst/sims/alertsim/avro_utils.py
Comment (danielsf) [inline python/lsst/sims/alertsim/avro_utils.py]: You need to update the docstring for this method to mention `session_dir`

File: python/lsst/sims/alertsim/avro_utils.py
Comment (danielsf) [inline python/lsst/sims/alertsim/avro_utils.py]: Just for future reference `os.path.join()` allows you to join an arbitrary number of strings into an intelligible path, i.e. `os.path.join('my','home','dir') == 'my/home/dir'`

File: python/lsst/sims/alertsim/avro_utils.py
Comment (danielsf) [inline python/lsst/sims/alertsim/avro_utils.py]: Can you not just call `json.dump(alert_dict, file_obj)`?

File: python/lsst/sims/alertsim/avro_utils.py
Comment (danielsf) [reply python/lsst/sims/alertsim/avro_utils.py]: Also, I think it would be best if you used `open()` as a context for this block of code.  For example:

```
with open('filename.txt', 'w') as out_file:
    for i in range(10):
        out_file.write('%d\n' % i)
```
writes 1...10 in `filename.txt` and automatically calls `close()` on the file, when you exit the `with` block.

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: There is now a method `lsst_camera` in `sims_coordUtils` that returns an LSST camera and keeps it in a global cache to prevent it from being re-instantiated needlessly.  You can use it like:
```
from lsst.sims.coordUtils import lsst_camera
camera = lsst_camera()
```

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: I think the first argument of `randomFloatArr` should be 20, not 32

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: Again: there aren't 32 columns in this getter

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: There appear to be 20 columns here, not 32

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: There appear to be 20 columns here, not 32

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: There appear to be 20 columns here, not 32

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: There appear to be 20 columns here, not 32

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: It does not look like this method is used.  It should be removed.

If I am wrong, you should have it raise an exception if neither `colname` nor `colname.lower()` are valid columns.

File: python/lsst/sims/alertsim/catalogs/diaSourceCommons.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/diaSourceCommons.py]: Can you either list the methods/classes being imported or import this with an alias so that we can tell where the imported methods and classes are coming from?  It will make it easier to maintain the code in the future.

File: python/lsst/sims/alertsim/catalogs/dia_transformations.py
Comment (danielsf) [inline python/lsst/sims/alertsim/catalogs/dia_transformations.py]: It does not look like this method is used.  You can delete it.

File: python/lsst/sims/alertsim/opsim_utils.py
Comment (danielsf) [inline python/lsst/sims/alertsim/opsim_utils.py]: You need to update the docstring of this method to reflect that `obs_gen` is no an argument.

**Review by danielsf — CHANGES_REQUESTED** [review]
I made some comments throughout, mostly about coding style.

The unit tests did not pass when I ran this branch on my machine, so you will need to fix that before you can merge.

I made a comment about the efficiency of generating the histories of DIAObjects in in 1-year intervals for each individual observation of an object.  That is not something that needs to be addressed in this Pull Request, but I do think that we need to talk about ways to speed up the performance of alertsim, now that we have a substantial number of variable sources in CatSim.

Let me know if you have any questions.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (danielsf) [reply python/lsst/sims/alertsim/alertsim_main.py]: Note the edit above.  You need to divide by cos(dec), not multiply

I guess that implies problems at the poles.  I'm not sure how we should deal with them.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: Couldn't get this to work - getting 
`AttributeError: module 'lsst.sims' has no attribute 'alertsim'`

However, this works fine - I guess it's good enough for now 
`from lsst.sims.alertsim import broadcast`

and even this
`from lsst.sims.alertsim import broadcast as bcast`

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (sambolino) [reply python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: sure, for all of them

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (sambolino) [reply python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: Will remove. This was a workaround at some point.

Comment (sambolino) [reply python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: also removed from diaSourceCommons.py

File: python/lsst/sims/alertsim/opsim_utils.py
Comment (sambolino) [reply python/lsst/sims/alertsim/opsim_utils.py]: done

File: python/lsst/sims/alertsim/catalogs/dia_transformations.py
Comment (sambolino) [reply python/lsst/sims/alertsim/catalogs/dia_transformations.py]: I think I'll need it, commented out and moved to bottom for now.

File: python/lsst/sims/alertsim/catalogs/diaSourceCommons.py
Comment (sambolino) [reply python/lsst/sims/alertsim/catalogs/diaSourceCommons.py]: done

File: python/lsst/sims/alertsim/catalogs/diaObjectCommons.py
Comment (sambolino) [reply python/lsst/sims/alertsim/catalogs/diaObjectCommons.py]: done, in diaSourceCommons too

File: python/lsst/sims/alertsim/avro_utils.py
Comment (sambolino) [reply python/lsst/sims/alertsim/avro_utils.py]: That wouldn't work earlier. Now looks fine but I ran into other issues - incompatibility of avro with py3 and json not willing to serialize numpy types. I think I managed to solve these, please look into it when I make a new pull req.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: Changed as you suggested, now alertId=diaSourceId, l1dbId=diaObjectId.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: Sure, this was only temporary. Will fail now on OSError, except when directory already exists. Dir name is timestamp and it's on the local machine, so there should never really be a problem. I can concatenate a random number as well.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: ok

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: You're right.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: Let's schedule a telecon on this and other issues after this merge

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: for now, eventid = alertId

File: python/lsst/sims/alertsim/avro_utils.py
Comment (sambolino) [reply python/lsst/sims/alertsim/avro_utils.py]: ok

File: python/lsst/sims/alertsim/avro_utils.py
Comment (sambolino) [reply python/lsst/sims/alertsim/avro_utils.py]: done

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (danielsf) [reply python/lsst/sims/alertsim/alertsim_main.py]: Sure

**Comment by danielsf** [comment]
Sorry I let this go unnoticed for so long.  Are you waiting for me to comment on some changes?

File: python/lsst/sims/alertsim/avro_utils.py
Comment (danielsf) [inline python/lsst/sims/alertsim/avro_utils.py]: You should also be able to use `DataFileWriter` as a context, i.e.
```
with DataFileWriter(open("avsc/alert.avro", "wb"), DatumWriter(), alert_schema) as writer:
    ...code to write data...
```
then you won't need to remember to call `writer.close()`

File: python/lsst/sims/alertsim/broadcast/receivers/rec_tcp.py
Comment (danielsf) [inline python/lsst/sims/alertsim/broadcast/receivers/rec_tcp.py]: I think you should also use `f` as a context, i.e
```
with open('VOEVENTs.txt', 'w') as f:
    ...code...
```

File: ups/sims_alertsim.table
Comment (danielsf) [inline ups/sims_alertsim.table]: I think the `if (type==exact)` form is automatically generated by EUPS at install time.  If you look at even the fundamental DM packages, for example:

https://github.com/lsst/afw/blob/master/ups/afw.table
https://github.com/lsst/obs_lsstSim/blob/master/ups/obs_lsstSim.table

the table files look more like what used to be in this file than what is in there now.

Is there a specific reason you made this change?

**Review by danielsf — APPROVED** [review]
In principle this is good to merge (assuming the unit tests pass).

I did have some in-line questions/suggestions that should be addressed.

File: python/lsst/sims/alertsim/broadcast/broadcast.py
Comment (sambolino) [reply python/lsst/sims/alertsim/broadcast/broadcast.py]: commented them out, will remove them eventually.

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: It's called only here, I grepped to make sure
grep -rnw python/ -e "load_parametrized_light_curves()"

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: sure, moved this and imports

File: python/lsst/sims/alertsim/avro_utils.py
Comment (sambolino) [reply python/lsst/sims/alertsim/avro_utils.py]: I'm a bit uncomfortable with so much nesting in that method

File: python/lsst/sims/alertsim/broadcast/receivers/rec_tcp.py
Comment (sambolino) [reply python/lsst/sims/alertsim/broadcast/receivers/rec_tcp.py]: sure, here's the ideal case because we're dealing with a stream

File: python/lsst/sims/alertsim/alertsim_main.py
Comment (sambolino) [reply python/lsst/sims/alertsim/alertsim_main.py]: I think it's more flexible to have try/catch here. Might handle some other failures in a specific manner if needed in future.

File: ups/sims_alertsim.table
Comment (sambolino) [reply ups/sims_alertsim.table]: No, that was my bad.

File: python/lsst/sims/alertsim/avro_utils.py
Comment (danielsf) [reply python/lsst/sims/alertsim/avro_utils.py]: Ok

**Comment by danielsf** [comment]
Go ahead and merge this whenever you are happy with it.

**Pull request merged**
